[DOC-10431] Further LTS improvements

mdlinville opened the pull request:
[DOC-10431] Further LTS improvements

1. Update the logic on the Releases index to get the LTS status from `site.data.versions` instead of `site.data.releases`
2. Update the version switcher to show LTS status both for the currently-viewed page version above the drop-down and in the list of versions in the drop-down
3. Modify the LTS Maintenance admonition from a red warning to a purple note.
4. Fix some LTS logic issues including off-by-one errors on LTS maintenance and assistance dates
5. In the Docker tab, hide the Notes column for all versions except for v23.1 and v22.2.
6. Fix a display bug each in the Source and Docker tabs

Also added some debug messages and the ability to enable or disable debug in the Releases index and the version switcher. The debug info shows in the page source as HTML comments.

### Previews
Description | URL | Description
------------|-----|-------------
Versioned page, not LTS, dev version | https://deploy-preview-18661--cockroachdb-docs.netlify.app/docs/v24.2/diagnostics-reporting (for example) |If the page exists for an LTS version, the entry for that version shows as LTS.
Versioned page, not LTS, stable version, in Assistance Support | https://deploy-preview-18661--cockroachdb-docs.netlify.app/docs/v24.1/diagnostics-reporting (for example) |If the page exists for an LTS version, the entry for that version shows as LTS. Red admonition.
Versioned page, LTS, in LTS Assistance Support | https://deploy-preview-18661--cockroachdb-docs.netlify.app/docs/v23.2/diagnostics-reporting (for example) |If the page exists for an LTS version, the entry for that version shows as LTS. Red admonition.
Versioned page, LTS, EOL | https://deploy-preview-18661--cockroachdb-docs.netlify.app/docs/v23.1/diagnostics-reporting (for example) | If the page exists for v23.1 and another version, v23.1 will be labeled as LTS and other entries do not. Red admonition.
Unversioned page (Release index)| https://deploy-preview-18661--cockroachdb-docs.netlify.app/docs/releases/. | LTS releases have (LTS) with a link to the Release Support Policy.
Unversioned page (Homepage) | https://deploy-preview-18661--cockroachdb-docs.netlify.app/docs/ | Unchanged by this PR.

To test the purple admonition for the terminal maintenance phase, you can set the date to 2024-11-13 and check a 23.1 version of a page.

ebembi-crdb reviewed (commented):
LGTM!
Nothing seems breaking, have we tested it?

mdlinville commented:
> LGTM! Nothing seems breaking, have we tested it?

Yes, I have tested it during implementation, but I'm hoping others will too. Some suggestions:

1. Validate that the Releases index still correctly shows LTS details now that it is looking to `versions.csv` rather than `release.yml`.
2. Validate that the version switcher correctly shows LTS status on versioned LTS pages.
3. Validate that the 23.1 LTS admonition now shows as a green tip instead of a red warning.
4. Validate that this doesn't break the functionality that was just rolled out in https://github.com/cockroachdb/docs/pull/18667 (23.1 should show the LTS admonition on all pages, not just the What's New page).

mikeCRL reviewed (changes requested):
- [ ] On the [Releases page](https://deploy-preview-18661--cockroachdb-docs.netlify.app/docs/releases/), all releases in the 23.2 table have "(LTS)" back through the first alpha.
- [ ] The dropdown shows (LTS) next to 24.1 and 24.2 when each is the selected version. I believe that is incorrect.
- [ ] I am seeing a red warning when viewing 24.1 and 23.2 pages (where the text correctly shows that maintenance support has ended but assistance is still active, per the 'today' date in the PR—LTS assistance in the case of 23.2). I believe these should now be purple?
- [ ] What case should be green, if any? How can I test that?

mdlinville commented on src/current/_includes/unsupported-version.md:
```diff
@@ -81,7 +81,7 @@ Today date: {{ today | date: "%Y-%m-%d" }} <br />
 
```
Done with a slight tweak. PTAL.

mikeCRL replied:
Nice. I like the clarity of showing the patch version.

mdlinville commented on src/current/_includes/unsupported-version.md:
```diff
@@ -105,7 +105,7 @@ Today date: {{ today | date: "%Y-%m-%d" }} <br />
 {% capture ga_maintenance_message %}
       {% if today <= m and today > a %}
-      {{site.data.alerts.callout_danger}}
+      {{site.data.alerts.callout_version}}
```
This is the one that shows up purple. Nothing shows up green now, I explain it below the screenshot in the description.

mdlinville commented:
> * [ ]  On the [Releases page](https://deploy-preview-18661--cockroachdb-docs.netlify.app/docs/releases/), all releases in the 23.2 table have "(LTS)" back through the first alpha.

Should be fixed.

> * [ ]  The dropdown shows (LTS) next to 24.1 and 24.2 when each is the selected version. I believe that is incorrect.

Should be fixed.

> * [ ]  I am seeing a red warning when viewing 24.1 and 23.2 pages (where the text correctly shows that maintenance support has ended but assistance is still active, per the 'today' date in the PR—LTS assistance in the case of 23.2). I believe these should now be purple?

Nope, the terminal phase is still red, as per Shannon.

> * [ ]  What case should be green, if any? How can I test that?

None should be green -- I mention this right below the screenshot showing green.

mdlinville commented:
The "Everything showing as LTS" bug should actually be fixed now, in this PR.

mdlinville commented on src/current/_includes/unsupported-version.md:
```diff
@@ -74,15 +74,15 @@ Today date: {{ today | date: "%Y-%m-%d" }} <br />
-      {{site.data.alerts.callout_danger}}
-      As of {{ x.lts_maint_supp_exp_date | date: "%B %e, %Y" }}, LTS releases of CockroachDB {{ x.major_version }} releases are in <strong>Maintenance Support</strong>{% if today > a %} and GA releases of CockroachDB {{ include.major_version }} are no longer supported.{% endif %}. Cockroach Labs will stop providing <strong>Assistance Support</strong> for {{ include.major_version }} LTS on <strong ...
+      {{site.data.alerts.callout_version}}
+      As of {{ x.lts_maint_supp_exp_date | date: "%B %e, %Y" }}, {{ include.major_version }} LTS releases of CockroachDB {{ x.major_version }} releases are in <strong>Maintenance Support</strong>{% if today > a %} and GA releases prior to {{ x.initial_lts_patch }} are no longer supported.{% endif %}. Cockroach Labs will stop providing <strong>Assistance Support</strong> for {{ include.major_versi ...
```
```suggestion
      As of {{ x.lts_maint_supp_exp_date | date: "%B %e, %Y" }}, LTS releases of CockroachDB {{ x.major_version }} are in <strong>Maintenance Support</strong>{% if today > a %} and GA releases prior to {{ x.initial_lts_patch }} are no longer supported.{% endif %}. Cockroach Labs will stop providing <strong>Assistance Support</strong> for {{ include.major_version }} LTS on <strong>{{ x.lts_asst_supp_exp_date | date: "%B %e, %Y" }}</strong>. Prior to that date, upgrade to a more recent version to continue receiving support. For more details, refer to the <a href="https://www.cockroachlabs.com/docs/releases/release-support-policy.html">Release Support Policy</a>
```

mdlinville commented on src/current/releases/index.md:
```diff
@@ -108,12 +133,29 @@ As of 2024, CockroachDB is released under a staged delivery process. New release
+                    {% endunless %}
+                {% endif %}
+
+                {% if DEBUG == true %}<tr><td colspan="3">current_patch: {{ current_patch }}<br />lts_patch: {{ lts_patch }}<br />r.release_name: {{ r.release_name }}<br />lts_link: {{ lts_link }}<br />in_lts: {{ in_lts }}</td>{% endif %}
```
This is not ideal but it just puts the debug info into a new row of the table.

mikeCRL reviewed (commented):
* Bug in the Source tab view. Not sure if I missed this on the previous review. https://deploy-preview-18661--cockroachdb-docs.netlify.app/docs/releases/?filters=source
* Docker tab issues https://deploy-preview-18661--cockroachdb-docs.netlify.app/docs/releases/?filters=docker ; generally the Notes column needs to be rechecked throughout.
  * For major versions with an LTS, note that LTS is displayed through the first non-experimental version.
  * For v23.1 Production:
    * Row with "Available only for select CockroachDB Cloud clusters" is missing a fourth cell.
    * LTS is listed back through 23.1.13 instead of 23.1.12
  * For Testing releases, can we leave out the GA/Experimental mentions? Perhaps leave it blank? Better to leave "Testing" be the only characterization than to use "GA" there, which is defined as production ready.

mdlinville commented:
> * Bug in the Source tab view. Not sure if I missed this on the previous review. https://deploy-preview-18661--cockroachdb-docs.netlify.app/docs/releases/?filters=source

There was some debug output present, now fixed.

> * Docker tab issues https://deploy-preview-18661--cockroachdb-docs.netlify.app/docs/releases/?filters=docker ; generally the Notes column needs to be rechecked throughout.
>   

It's sad, but we need to redo the looping logic from the Linux tab in the Mac tab. Fixed. Also moved the "LTS" over to the first column to align with the Linux tab.

The reason the Notes column is there is because in v22.2 and v23.1, the contents of all rows are not the same. I will need to conditionalize the # of columns based on the version number to remove them. It's do-able, just adds a bit more complexity to the upkeep of the page.

Let me summarize what I did to fix the Docker tab:
1. Conditionalize the intro para to show only before the Production section, not for testing releases.
2. In the intro text, for versions where all Docker images are multi-platform, state that. For mixed versions, retain the explanation about Intel vs multi-platform. For pre-ARM versions, 
3. Show the Notes column only for versions with both GA and pre-GA ARM support (v22.2 and v23.1)
4. In the Notes column, don't print Intel because it's implied and also covered in the Docker intro.

>   * For major versions with an LTS, note that LTS is displayed through the first non-experimental version.
Fixed. I needed to re-initialize some variables in the Docker loop.

>   * For v23.1 Production:
>     
>     * Row with "Available only for select CockroachDB Cloud clusters" is missing a fourth cell.
It's actually not, it's a colspan. I tested locally after designating 23.1.24 as Cloud-only:

![Screenshot 2024-07-31 at 2 05 42 PM](https://github.com/user-attachments/assets/45c5214d-a18e-45b6-9d71-5d9af4ac25c0)

>     * LTS is listed back through 23.1.13 instead of 23.1.12

I can't reproduce this. Verified in versions.csv that 23.1.11 is listed as the last GA patch and 23.1.12 as the first LTS patch, and here is how it looks with a local build:
![Screenshot 2024-07-31 at 2 08 46 PM](https://github.com/user-attachments/assets/7f3cb9c1-9b7d-4072-9f68-9ddd556d8dcf)

>   * For Testing releases, can we leave out the GA/Experimental mentions? Perhaps leave it blank? Better to leave "Testing" be the only characterization than to use "GA" there, which is defined as production ready.

I've minimized it, but I am pretty sure I should not remove Experimental from the versions for which ARM actually was Experimental. It's only still listed in v22.2 and v23.1, which had a mix. We need to leave that in there so that customers don't think ARM was supported prior to v22.2.17, v23.1.12, or v23.2.0, including the testing releases leading up to v23.2.0.

mikeCRL reviewed (approved):
LGTM
